transformer_engine/jax/fp8.py

```diff
@@ -82,72 +83,67 @@ class FP8MetaPackage:
     A container that contains all required meta data for FP8
     """
 
+    NUM_OF_META: int = 3
+    INPUT_IDX: int = 0
+    WEIGHT_IDX: int = 1
+    GRAD_IDX: int = 2
+
     def __init__(
         self,
-        num_of_gemm: int,
-        fp8_max: jnp.ndarray,
-        amax: jnp.ndarray,
-        scale: jnp.ndarray,
-        scale_inv: jnp.ndarray,
+        i_amax: jnp.ndarray,
+        i_scale: jnp.ndarray,
+        w_amax: jnp.ndarray,
+        w_scale: jnp.ndarray,
+        g_amax: jnp.ndarray,
+        g_scale: jnp.ndarray,
     ) -> None:
-        total_num_of_meta = num_of_gemm * FP8Helper.NUM_META_PER_GEMM
-        self._num_of_gemm = num_of_gemm
-        assert fp8_max.shape[0] == total_num_of_meta
-        self._fp8_max = fp8_max
-        assert amax.shape[0] == total_num_of_meta
-        self._amax = amax
-        assert scale.shape[0] == total_num_of_meta
-        self._scale = scale
-        assert scale_inv.shape[0] == total_num_of_meta
-        self._scale_inv = scale_inv
 
-    @property
-    def num_of_gemm(self) -> int:
-        """
-        num_of_gemm of this package
-        """
-        return self._num_of_gemm
+        self._amax_list = [None] * FP8MetaPackage.NUM_OF_META
+        self._scale_list = [None] * FP8MetaPackage.NUM_OF_META
 
-    @property
-    def fp8_max(self) -> jnp.ndarray:
-        """
-        fp8_max of this package
-        """
-        return self._fp8_max
+        self._amax_list[FP8MetaPackage.INPUT_IDX] = i_amax
+        self._scale_list[FP8MetaPackage.INPUT_IDX] = i_scale
+        self._amax_list[FP8MetaPackage.WEIGHT_IDX] = w_amax
+        self._scale_list[FP8MetaPackage.WEIGHT_IDX] = w_scale
+        self._amax_list[FP8MetaPackage.GRAD_IDX] = g_amax
+        self._scale_list[FP8MetaPackage.GRAD_IDX] = g_scale
```

Comment:
I'm wondering if `dict` a better structure for maintaining the FP8 meta, we can use `jax.tree` for FP8 meta operations.

Reply:
The `list` and `dict` seem to have the simialr advantages here, since we eventially have to slice those meta to `amax` and `scale`. The difference here when to do so. If the structure is `list` then we split meta before calling VJP functions. If it is `dict`, then we could delay the split into VJP functions.

For `jax.tree`, since there are only one function could be shared between `amax` and `scale`, that is `converting to/from fm32`, then there are not much changes to take to advantages of `jax.tree`.

I think it might be fine to make it as `list` to be more straightforward at this moment, but we could discuss more offline.

Reply:
It seems there is no special reason that the `dict` outperform `list` besides my personal preference. Let's just keep it as list. By the way, `jax.tree.map` also works on `list`, we can use that next time.